grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java

```diff
@@ -416,6 +429,53 @@ private static Descriptor findTypeDescriptor(FileDescriptor file, FieldDescripto
         return file.findMessageTypeByName(messageType.getName());
     }
 
+    // to make it more efficient, we calculate whether extract response body one time
+    // if there is no matching toplevel field, we set it to null
+    @Nullable
+    private static String calculateResponseBody(List<FieldDescriptor> topLevelFields, String responseBody) {
+        if (StringUtil.isNullOrEmpty(responseBody)) {
+            return null;
+        }
+        for (FieldDescriptor fieldDescriptor: topLevelFields) {
+            if (fieldDescriptor.getName().equals(responseBody)) {
+                return responseBody;
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private static Function<HttpData, HttpData> generateResponseBodyConverter(TranscodingSpec spec) {
+        final String responseBody = spec.httpRule.getResponseBody();
+        if (spec.responseBody == null) {
+            return null;
+        } else {
+            return httpData -> {
+                try (HttpData body = httpData) {
+                    final byte[] array = body.array();
+                    try {
+                        final JsonNode jsonNode = JacksonUtil.readValue(array, JsonNode.class);
+                        final String responseBodyJsonKey =
+                                CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.LOWER_CAMEL, responseBody);
+                        final Iterator<String> fieldNames = jsonNode.fieldNames();
+                        while (fieldNames.hasNext()) {
+                            final String fieldName = fieldNames.next();
+                            if (CaseFormat.UPPER_UNDERSCORE.to(
+                                    CaseFormat.LOWER_CAMEL, responseBody).equalsIgnoreCase(fieldName)) {
+                                final JsonNode responseBodyJsonNode = jsonNode.get(responseBodyJsonKey);
+                                final byte[] bytes = JacksonUtil.writeValueAsBytes(responseBodyJsonNode);
+                                return HttpData.copyOf(bytes);
+                            }
+                        }
+                        return HttpData.ofAscii("null");
+                    } catch (IOException e) {
+                        return HttpData.ofAscii("null");
```

Comment:
How about returning the original array if there is no matching field or an exception is raised?

Reply:
You're right.

Reply:
@ikhoon When there is no matching field when iterating the json node, since in the previous `calculateResponseBody`  method already decide the response body matches a top-level field in protobuf, it maybe because the incoming request does not have that field. So I think maybe returning null is a better solution when there is no matching field. But when exception is raised returning original data is better. Thanks